From 5a8e3275ff315f1a623600f58b83ed1451827fcd Mon Sep 17 00:00:00 2001
From: alvindimas05 <alvin.dimas.praditya@gmail.com>
Date: Fri, 23 Oct 2026 01:52:33 +0700
Subject: [PATCH 19/33] [texture_cache] single-page fastpath for
 ForEachImageInRegion CPU and GPU

(cherry picked from commit aef4efa4c87602987e686639e310084a8a43bc93)
---
 src/video_core/texture_cache/texture_cache.h | 51 ++++++++++++++++++++
 1 file changed, 51 insertions(+)

diff --git a/src/video_core/texture_cache/texture_cache.h b/src/video_core/texture_cache/texture_cache.h
index 4e5ffb193..042abbd35 100644
--- a/src/video_core/texture_cache/texture_cache.h
+++ b/src/video_core/texture_cache/texture_cache.h
@@ -2007,6 +2007,32 @@ template <typename Func>
 void TextureCache<P>::ForEachImageInRegion(DAddr cpu_addr, size_t size, Func&& func) {
     using FuncReturn = typename std::invoke_result<Func, ImageId, Image&>::type;
     constexpr bool BOOL_BREAK = std::is_same_v<FuncReturn, bool>;
+    if (size == 0) [[unlikely]] {
+        return;
+    }
+    const u64 page_start = cpu_addr >> YUZU_PAGEBITS;
+    const u64 page_end = (cpu_addr + size - 1) >> YUZU_PAGEBITS;
+    if (page_start == page_end) {
+        const auto it = page_table.find(page_start);
+        if (it == page_table.end()) {
+            return;
+        }
+        for (const ImageMapId map_id : it->second) {
+            ImageMapView& map = slot_map_views[map_id];
+            if (!map.Overlaps(cpu_addr, size)) {
+                continue;
+            }
+            Image& image = slot_images[map.image_id];
+            if constexpr (BOOL_BREAK) {
+                if (func(map.image_id, image)) {
+                    return;
+                }
+            } else {
+                func(map.image_id, image);
+            }
+        }
+        return;
+    }
     boost::container::small_vector<ImageId, 32> images;
     boost::container::small_vector<ImageMapId, 32> maps;
     ForEachCPUPage(cpu_addr, size, [this, &images, &maps, cpu_addr, size, func](u64 page) {
@@ -2066,6 +2092,31 @@ void TextureCache<P>::ForEachImageInRegionGPU(size_t as_id, GPUVAddr gpu_addr, s
         return;
     }
     auto& gpu_page_table = gpu_page_table_storage[*storage_id * 2];
+    if (size == 0) [[unlikely]] {
+        return;
+    }
+    const u64 page_start = gpu_addr >> YUZU_PAGEBITS;
+    const u64 page_end = (gpu_addr + size - 1) >> YUZU_PAGEBITS;
+    if (page_start == page_end) {
+        const auto it = gpu_page_table.find(page_start);
+        if (it == gpu_page_table.end()) {
+            return;
+        }
+        for (const ImageId image_id : it->second) {
+            Image& image = slot_images[image_id];
+            if (!image.OverlapsGPU(gpu_addr, size)) {
+                continue;
+            }
+            if constexpr (BOOL_BREAK) {
+                if (func(image_id, image)) {
+                    return;
+                }
+            } else {
+                func(image_id, image);
+            }
+        }
+        return;
+    }
     ForEachGPUPage(gpu_addr, size,
                    [this, &gpu_page_table, &images, gpu_addr, size, func](u64 page) {
                        const auto it = gpu_page_table.find(page);
-- 
2.43.0

